QUICK-123 Added HardwareApdexAnalysisIT to allow exploring discrepanc…

mminns opened the pull request:
…ies between apdex over runs on the same hardware and size.


First Kotlin PR so do your worst :)

mminns committed:
QUICK-123 Added HardwareApdexAnalysisIT to allow exploring discrepancies between apdex over runs on the same hardware and size.

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+
+    // The following needs to match the configuration of the test run you want to analyze
+    // Taken from extraLarge.Kt hold + ramp + flat values
+    private val testDuration: Duration = Duration.ZERO + Duration.ofSeconds(90) + Duration.ofMinutes(20)
```
we can actually call `extraLarge(false, false).load.total`

mminns replied:
Yes I did try that and then consciously went back to this, since using extraLarge() requires a Jira Licence, which seemed overkill for this kind of test/process

```
java.lang.Exception: Put a Jira license to /Volumes/PROJECTS/projects/github.com/mminns/jira-hardware-exploration/jira-license.txt

	at com.atlassian.performance.tools.hardware.HwrDatasetCatalogue.overrideLicense(HwrDatasetCatalogue.kt:162)
```

dagguh replied:
Ooooh, I see. Ok

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/lib/table/GenericPlainTextTable.kt:
```diff
@@ -0,0 +1,102 @@
+ *  table.addRow(listOf( "", "Run", "Apdex", "Total Actions", "Error %", "Satisfactory %", "Tolerable %"))
+ *  table.addRow(listOf( "best", "5", "0.7713201820940819", "23155", "0.389", "59.490", "34.684"))
+ *  table.addRow(listOf( "worst", "6", "0.7677944046844503", "23143", "0.380", "58.925", "35.125"))
+ *  println(table.generate())
```
This looks like a unit test already 🙂 
Unit tests have the advantage of automatically validating the usage and creating a simple environment to tinker, debug, adapt

mminns replied:
Not sure I follow this comment.

This class is just a reporting tool a re-usable way to write out ascii tables. It is used by the Unit Test, but doesn't contain any test logic.

dagguh replied:
The comment has some value.
A `GenericPlainTextTableTest` would have the same value plus more. It would be cheap to impl, because we can use the code from the comment.

mminns replied:
Ah I understand.

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/lib/s3cache/S3Cache.kt:
```diff
@@ -56,7 +60,9 @@ class S3Cache(
         val cachedEtag = etags.read(objectSummary.key)
+        val match = matchesSearchPattern(objectSummary)
         return when {
+            !match -> false
```
Nice one!

mminns replied:
Yeah that made my kotlin-newbie head spin a bit, glad I used linq a lot in c# :)

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/lib/WeightedActionMetric.kt:
```diff
@@ -0,0 +1,8 @@
+import com.atlassian.performance.tools.jiraactions.api.ActionMetric
+
+/**
+ * Wrapper for ActionMetric which can store the apdex 'weight' applied alongside this metric
```
This can be easily confused with [synthetic Apdex weighing](https://hello.atlassian.net/wiki/spaces/DC/pages/353104296/Hardware+recommendations+-+Confluence+Large+profile?focusedCommentId=362259099#comment-362259099) we used for Confluence.
`Apdex` has a `score` method, perhaps this could be called `ScoredActionMetric` or `ApdexActionMetric`?

PS. In the partial (paused and not merged) bottleneck analysis I also needed the Apdex score classification and it looks like this: 1035cf16620d4e3a7d220e01f103e638864ac724
Feel free to reuse or ignore

mminns replied:
I'll rename to avoid confusion.

re: https://github.com/atlassian/jira-hardware-exploration/commit/1035cf16620d4e3a7d220e01f103e638864ac724 I think there is overlap but i think they are complimentary rather than equivalent, but I'll take a look.

mminns replied:
Having reviewed my code further I do think there is overlap with https://github.com/atlassian/jira-hardware-exploration/commit/1035cf16620d4e3a7d220e01f103e638864ac724

i.e. later in my code I want counts of Satisfactory, Frustrated etc metrics and effectively run the filters you code into  Apdex.kt

However I want to keep the individual scores for additional calculations so I think both add useful functionality to Apdex.kt

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+                    ApdexResults(run.toInt(), apdex, weightedMetrics)
+
+                } else {
+                    println("ignominious failure! ${failure.localizedMessage}")
```
😆

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/lib/table/GenericPlainTextTable.kt:
```diff
@@ -0,0 +1,102 @@
+ * Assumes the first row added is the header, but that can be defined in the constructor. Dynamically works out column widths so you don't have to.
+ */
+class GenericPlainTextTable (private val headerRowCount: Int = 1) {
+    private val _builder = StringBuilder()
```
No need for `_`, because the `private` keyword does the job

mminns replied:
~~I'll edit it~~

Actually just to check the yes the `private` defines the scope the _ is just for naming, to highlight out the use of the private variables elsewhere in the code.

I'm not religiously tied to that nomenclature, but does it go against any code conventions?

dagguh replied:
It doesn't directly go against the [Kotlin Coding Conventions](https://kotlinlang.org/docs/reference/coding-conventions.html#naming-rules)

dagguh commented:
Good stuff @mminns 👍 
I reviewed a part of it and I'll finish it tomorrow

kema2000 commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+            }
+            .filter { file ->
+                // ignore any empty run/# folders or ones that only contain a status.txt which indicates a failed run
+                file.listFiles().any{ it.name != "status.txt" }
```
There are some cases the status.txt is generated with error message. E.g. one of the virtual user fails, you will still be able to find some action-matrics file.
I think you might want to filter out those file by checking if the content of the status.txt is "Ok" or not.
/com/atlassian/performance/tools/lib/Report.kt should have the logic already implemented.

kema2000 replied:
you actually did filter out those result inside apdexMetrics() :)

mminns commented:
Ah, I was binning all my GitHub.com notifications so i missed your comments. I'll take a look now I've seen them :)

atl-mk commented:
@dagguh whoops, soz mistapped when reading and there's no undo

dagguh commented:
https://bulldog.internal.atlassian.com/browse/QUICK-123 is done. Is this PR still a part of QUICK-123 or some other issue or abandoned?

mminns commented:
It isn't abandoned, I just missed the updates due to a badly configured GMail filter, but I am open to advice/opinions on whether it should be merged or not.

Should the jira-hardware-exploration contain all of the code used during the work, even if they may not be run again in this form?

If so then yes it I would like to see it merged.

If not I do think there a some aspects that are worth preserving, i.e. 
https://github.com/atlassian/jira-hardware-exploration/pull/40/files#diff-9abbd31ec4f9ac788917938a7f8bd6b5 <- possibly moving to https://github.com/atlassian/jira-actions ?
https://github.com/atlassian/jira-hardware-exploration/pull/40/files#diff-a3bb3210cec35644ac8cf0da37fbc029 
https://github.com/atlassian/jira-hardware-exploration/pull/40/files#diff-587e037a6873e0cce43b8f11dd0fb2a6 <- possibly moving to https://github.com/atlassian/report ?

Any thouhts?

dagguh commented:
The PR aims to provide reproducible analysis to answer the QUICK-123 question.
If it does so, then it's worth merging.

Unfortunately, when I run it myself I get this:
```
14:02:54,818 DEBUG [download, filter] The task took PT40.985S
14:02:54,820 INFO  [download, transfer] Downloading 0, skipping 67191
14:02:54,821 DEBUG [download, transfer] The task took PT0.001S
14:02:54,822 DEBUG [download] The task took PT40.989S

java.util.NoSuchElementException: Collection contains no element matching the predicate.
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT.apdexMetrics(HardwareApdexAnalysisIT.kt:356)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT.access$apdexMetrics(HardwareApdexAnalysisIT.kt:31)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT$shouldAnalyzeAppdex$runs$3.invoke(HardwareApdexAnalysisIT.kt:63)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT$shouldAnalyzeAppdex$runs$3.invoke(HardwareApdexAnalysisIT.kt:31)
	at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:149)
	at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:647)
	at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:677)
	at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:668)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT.shouldAnalyzeAppdex(HardwareApdexAnalysisIT.kt:65)
```

mminns commented:
Interesting I'll clean my cached data and try re-running here.

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+
+    private val logger: Logger = logContext.getLogger(this::class.java.canonicalName)
+
+    // define the appdex sets we want to investigate
```
```suggestion
    // define the Apdex sets we want to investigate
```

mminns replied:
done 👍

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+
+    // define the appdex sets we want to investigate
+    private val jiraInstanceTypeOfInterest: InstanceType = C518xlarge
+    private val nodeCountOfInterest: Int = 7
```
We can define the entire `hardwareOfInterest: Hardware`
It also contains `db: InstanceType`
Did we skip DB filtering by design or by accident?

mminns replied:
By Accident I think, I will add that in.

mminns replied:
done 👍

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+    fun shouldAnalyzeAppdex() {
+
+        // https://docs.oracle.com/javase/tutorial/essential/io/fileOps.html#glob
+        // action-metrics.jpt contain the appdex information
```
```suggestion
        // action-metrics.jpt contain the Apdex information
```

mminns replied:
done 👍

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+                    .asSequence().singleOrNull {
+                        it.key == entry.key
+                    }
+                if(worstRow != null) {
```
There are small code-style inconsistencies here and there.
I recommend spamming IntelliJ `Reformat Code` action. I do it all the time.
This way it is not only easier on the eyes, but we're avoiding generating unnecessary diff in future commits.

mminns replied:
done 👍

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,269 @@
+                }
+            }
+            // this seems very hacky, I should just be able to get 1 result out of the sequence code above ... ?
+            .first{ it is ApdexResults}
```
This tries to detect whether we `map`ped to an `ApdexResults` or a `null`.
We can do it directly, by changing the `map` above to a `mapNotNull` and then `first()`.

PS. We can extract the `map` contents to a private method to see the full chain at once.

mminns replied:
done 👍

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/lib/Apdex.kt:
```diff
@@ -12,17 +12,30 @@ class Apdex {
+            .filter { it.result == ActionResult.OK }))
+    }
+
+    fun average(
```
`score` filters by `OK`, but `average` and `weight` don't. This can easily lead to non-obvious interpretations further down the pipeline. I'd keep the consistency one way or another.

Taking a step back - what *should* be the Apdex relation to errors? Apdex is a measure of user experience. In my book, getting an error is definitely not a _satisfactory_ experience and probably not even _tolerable_. It's _frustrating_.

We can filter errors out and rely on other metrics to punish errors (e.g. error rate).
We can also leave errors in and make Apdex punish errors by itself.

dagguh replied:
PS. This is also why it's worth importing parts of https://github.com/atlassian/jira-hardware-exploration/commit/1035cf16620d4e3a7d220e01f103e638864ac724, especially `enum class Experience`, because it reminds that Apdex is all about UX.

mminns replied:
I've pulled in the Experience, but I think the definition of Apdex is a wider question :)

mminns replied:
I have pulled on the Experience, but while I agree with you I think changing the Apdex definition is beyond this scope.

I have not changed the ActionResult filter because the text explicitly needs unfiltered metrics, but I have renamed the functions to hopefully be more explicit.

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/lib/Apdex.kt:
```diff
@@ -12,17 +12,30 @@ class Apdex {
             .average()
     }
 
+    fun weight(
```
In general, functions should be named after activities (verb-like), so we'd go for `weigh`

mminns replied:
done 👍

mminns commented:
I suspect the crash you are seeing is because of the file path filter not being as specific as it should be. You have a lot more locally cached data than I do, I'll look at it tomorrow.

mminns commented:
Yes the crash is because of more files/folders in your environment. I will fix it up

mminns committed:
QUICK-123 PR changes
Rationalize Apdex.kt to make it clearer about the purpose of its functions.
Refactored test code to make it more readable
Formatted code
Use a Hardware definition for Jira, DB instance types and node count definitions.
Correct Apdex spelling
Rename the 'weighted' Apdex to 'Scored' Apdex
Added a test to act as documentation for the Generic Table code

mminns commented:
I've applied a better search pattern which should exclude everything other than the explicit data we are interested in

dagguh commented:
It still fails with:
```
java.util.NoSuchElementException: List is empty.

	at kotlin.collections.CollectionsKt___CollectionsKt.first(_Collections.kt:184)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT.apdexMetrics(HardwareApdexAnalysisIT.kt:193)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT.access$apdexMetrics(HardwareApdexAnalysisIT.kt:31)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT$shouldAnalyzeApdex$runs$3.invoke(HardwareApdexAnalysisIT.kt:66)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT$shouldAnalyzeApdex$runs$3.invoke(HardwareApdexAnalysisIT.kt:31)
	at kotlin.sequences.TransformingSequence$iterator$1.next(Sequences.kt:149)
	at kotlin.sequences.SequencesKt___SequencesKt.toCollection(_Sequences.kt:647)
	at kotlin.sequences.SequencesKt___SequencesKt.toMutableList(_Sequences.kt:677)
	at kotlin.sequences.SequencesKt___SequencesKt.toList(_Sequences.kt:668)
	at com.atlassian.performance.tools.hardware.HardwareApdexAnalysisIT.shouldAnalyzeApdex(HardwareApdexAnalysisIT.kt:68)
```

dagguh commented on src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt:
```diff
@@ -0,0 +1,273 @@
+            }
+            .first()
+
+        return cohortResults!!
```
IntelliJ warns:
![image](https://user-images.githubusercontent.com/1194704/59277916-50fd3a80-8c61-11e9-999d-9af7935ae2ac.png)

mminns replied:
will tidy that up.

mminns replied:
Do you know if IntelliJ allows you to make the analysis warnings into build failures?

dagguh replied:
It's a compiler-emitted warning
We can use this:
```
tasks.withType<KotlinCompile> {
    kotlinOptions.allWarningsAsErrors = true
}
```
Perhaps we'll finally fix the `kotlin-stdlib-jre7` deprecation warnings: https://ecosystem.atlassian.net/browse/JPERF-466

mminns commented:
Not quite as resilient as I hoped then ;)

mminns committed:
QUICK-123 warning fixes

mminns committed:
QUICK-123 further hardening.
One oddity is that, I believe the .none(), effectively means the sequence is run 1.5 times. The result is fine, but it effects performance.

mminns commented:
@dagguh when you have some time could you try running this again? I added some additional hardening.

dagguh commented:
On it

dagguh commented:
Heey it works!
```
*------*----*-------------------*--------------*--------*---------------*------------*
|      | Run| Apdex             | Total Actions| Error %| Satisfactory %| Tolerable %|
*------*----*-------------------*--------------*--------*---------------*------------*
| best | 5  | 0.7713201820940819| 23155        | 0.389  | 59.490        | 34.684     |
| worst| 6  | 0.7677944046844503| 23143        | 0.380  | 58.925        | 35.125     |
*------*----*-------------------*--------------*--------*---------------*------------*
```

mminns commented:
Happy for me to merge?

dagguh commented:
LGTM let's ship it :)

dagguh commented:
Ah snap, the master has moved significantly forward since the last pushes to this branch. And after rebasing, this doesn't compile anymore.
I'll attach a simple CI to at least cover the compilation

dagguh commented:
CircleCI added in https://github.com/atlassian/jira-hardware-exploration/pull/54
